src/Fabulous/Builders.fs

```diff
@@ -13,49 +12,50 @@ type AttributesBundle =
 [<Struct; NoComparison; NoEquality>]
 type WidgetBuilder<'msg, 'marker> =
     struct
-        val Key: WidgetKey
+        val Definition: WidgetDefinition
         val Attributes: AttributesBundle
 
-        new(key: WidgetKey, attributes: AttributesBundle) = { Key = key; Attributes = attributes }
+        new(definition: WidgetDefinition, attributes: AttributesBundle) =
+            { Definition = definition
+              Attributes = attributes }
 
-        new(key: WidgetKey, scalar: ScalarAttribute) =
-            { Key = key
+        new(definition: WidgetDefinition, scalar: ScalarAttribute) =
+            { Definition = definition
               Attributes = AttributesBundle(StackList.one scalar, ValueNone, ValueNone) }
 
-        new(key: WidgetKey, scalarA: ScalarAttribute, scalarB: ScalarAttribute) =
-            { Key = key
+        new(definition: WidgetDefinition, scalarA: ScalarAttribute, scalarB: ScalarAttribute) =
+            { Definition = definition
               Attributes = AttributesBundle(StackList.two (scalarA, scalarB), ValueNone, ValueNone) }
 
-        new(key: WidgetKey, scalar1: ScalarAttribute, scalar2: ScalarAttribute, scalar3: ScalarAttribute) =
-            { Key = key
+        new(definition: WidgetDefinition, scalar1: ScalarAttribute, scalar2: ScalarAttribute, scalar3: ScalarAttribute) =
+            { Definition = definition
               Attributes = AttributesBundle(StackList.three (scalar1, scalar2, scalar3), ValueNone, ValueNone) }
 
         [<EditorBrowsable(EditorBrowsableState.Never)>]
         member x.Compile() : Widget =
             let struct (scalarAttributes, widgetAttributes, widgetCollectionAttributes) = x.Attributes
 
-            { Key = x.Key
-#if DEBUG
-              DebugName = $"{typeof<'marker>.Name}<{typeof<'msg>.Name}>"
-#endif
-              ScalarAttributes =
-                  match StackList.length &scalarAttributes with
-                  | 0us -> ValueNone
-                  | _ -> ValueSome(Array.sortInPlace (fun a -> a.Key) (StackList.toArray &scalarAttributes))
-
-              WidgetAttributes = ValueOption.map (Array.sortInPlace (fun a -> a.Key)) widgetAttributes
-
-
-              WidgetCollectionAttributes =
-                  widgetCollectionAttributes
-                  |> ValueOption.map (Array.sortInPlace (fun a -> a.Key)) }
+            { Definition = x.Definition
+              Data =
+                  { ScalarAttributes =
+                      match StackList.length &scalarAttributes with
+                      | 0us -> ValueNone
+                      | _ -> ValueSome(Array.sortInPlace (fun a -> a.Definition.Key) (StackList.toArray &scalarAttributes))
```

Comment:
That is what I was referring when I mentioned that we need a fast comparison key, unfortunately `a.Definition.Key` is a potential double cache miss. And `string` is heap allocated value itself, thus comparing it will involve another memory jump.

So the way I was thinking about it is to use `string` as a user facing `Key`: requite unique string for each widget and for each attribute (exactly as you do it currently here). But with a twist:
1. When you create a definition calculate a hash of that string and store it in the attribute and addition to the reference to the definition. `{Key = "AttrKey"; Hash = 14565; ...}`
2. in DEBUG mode have a global table lookup to ensure uniqueness (both for string and hash)

Then we have direct access to Definition but also have a key for fast sorting (int)

I was planning to use that `hash` prop to store additional attributes in a couple of bits.